### PR TITLE
ID-347 Remove Martha and use DRSHub in tests

### PR DESCRIPTION
**Pull request by andy7i**
What:
Adds configuration options for Leonardo to use drshub instead of martha in unit tests.
https://broadworkbench.atlassian.net/browse/ID-347

Why:
So we can test switching over to using DrShub instead of Martha.

How:
Updating unit tests.

---
Have you read [CONTRIBUTING.md](https://github.com/DataBiosphere/leonardo/blob/develop/CONTRIBUTING.md) lately? If not, do that first.

I, the developer opening this PR, do solemnly pinky swear that:

- [ ] I've documented my API changes in Swagger

In all cases:

- [ ] Get a thumbsworth of review and PO signoff if necessary
- [ ] Verify all tests go green (you can re-run automation tests with a comment saying `jenkins retest`
- [ ] Run the automation tests multiple times in parallel to weed out instability if applicable via a comment saying `multi-test`
- [ ] Squash and merge; Delete your branch after this
- [ ] Test this change deployed correctly and works on dev environment after deployment

### CODE REVIEW

File: http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/config/Config.scala
Comment (Qi77Qi) [inline http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/config/Config.scala]: any reason we can't switch to drshub yet?

Comment (andy7i) [reply http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/config/Config.scala]: good question. @tlangs do you know the answer to this?

Comment (andy7i) [reply http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/config/Config.scala]: the ticket mentions "Just configuration options that aren't utilized yet"

Comment (tlangs) [reply http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/config/Config.scala:612]: Yeah. We haven't performance-tested DRSHub yet, and making things talk to it without first doing that seems unwise. The plan is to test it sequentially in all pre-prod envs to make sure nothing breaks, then performance test it  to make sure it can at least match Martha, then push to prod.

File: http/src/main/resources/reference.conf
Comment (tlangs) [inline http/src/main/resources/reference.conf]: This needs to be `https://drshub.dsde-dev.broadinstitute.org/api/v4/drs/resolve`

File: http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/config/Config.scala
Comment (tlangs) [reply http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/config/Config.scala]: Yeah. We want to make sure DRSHub can match Martha's performance before going to prod. The plan is to configure all services to be able to talk to DRSHub, then flip the switches in non-prod envs sequentially and see what happens. We'll want to perf test it before going to prod as well. So for now, things should continue using Martha.

**Review by tlangs — CHANGES_REQUESTED** [review]

**Review by tlangs — APPROVED** [review]

**Review by Qi77Qi — COMMENTED** [review]
is related firecloud-develop PR merged?

File: http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/config/Config.scala
Comment (Qi77Qi) [reply http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/config/Config.scala]: I probably would've just used different value for different envs instead of introducing 2 config values.

So basically just adding `drshubUrl`, and the value for it will vary depending on the env, which will be a firecloud-develop PR

File: http/src/main/resources/reference.conf
Comment (rtitle) [inline http/src/main/resources/reference.conf]: Noting that this URL actually gets sent to Galaxy, and Galaxy makes DRS calls from python. I think this is the code: https://github.com/anvilproject/fs.anvilfs

Does the DRSHub API have the same shape as martha? E.g. same auth, params, etc. If so then it should be safe, but ideally we'd test with Galaxy before cutting over.

**Review by rtitle — COMMENTED** [review]
This looks fine to me but I defer to @Qi77Qi. :) I agree fc-develop needs to merge first because this introduces a new environmental config value. I left a comment on the fc-develop PR about BEE-ificiation.

File: http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/config/Config.scala
Comment (tlangs) [reply http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/config/Config.scala]: So, I'm hearing you'd rather have the rendered environment determine the `drsUrl`, rather than doing it in code?

Comment (Qi77Qi) [reply http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/config/Config.scala]: yea...no big deal...they do the same thing...just feels the current approach is slightly more complicated than necessary

When later we do the full switch, it'd be easier to just change the config value in firecloud-develop too instead of having to change both leo and firecloud-develop

File: http/src/main/resources/reference.conf
Comment (Qi77Qi) [reply http/src/main/resources/reference.conf]: @tlangs @andy7i can u comment on this..I thikn this is the most important thing to confirm besides making sure `fireloud-develop` is merged before this one

**Comment by andy7i** [comment]
As noted in the PR description, I'll wait to merge this PR until after the fc-dev PR has merged.

File: http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/config/Config.scala
Comment (andy7i) [reply http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/config/Config.scala]: Going with this approach, I'm realizing that the unit tests can either test Martha or DRSHub but not both.  I'm wondering if we should leave the Martha code or if we want to switch over tests to DRSHub now. I think could keep the Martha code until we're ready to release the DRSHub code. No strong preference.

File: http/src/main/resources/reference.conf
Comment (andy7i) [reply http/src/main/resources/reference.conf]: based on the documentation, it looks very similar. 
https://drshub.dsde-dev.broadinstitute.org/#/drsHub/resolveDrs
https://github.com/broadinstitute/martha#martha-v3

@tlangs do you know off the top of your head whether the two APIs are functionally equivalent (for example, parameters, response, etc.)?

**Comment by andy7i** [comment]
jenkins retest - startup failed

```
10:51:38 3:51PM ERR 1 execution errors:
10:51:38 jobmanager: Command "ARGOCD_SERVER=ap-argocd.dsp-devops.broadinstitute.org ARGOCD_AUTH_TOKEN=****** argocd --header Proxy-Authorization: Bearer ****** --grpc-web app get jobmanager-jenkins-swat-leo-5446 -o yaml" exited with status 20:
10:51:38 time="2022-11-17T15:48:08Z" level=fatal msg="rpc error: code = Unknown desc = Post \"[https://ap-argocd.dsp-devops.broadinstitute.org:443/application.ApplicationService/Get\](https://ap-argocd.dsp-devops.broadinstitute.org/application.ApplicationService/Get/)": dial tcp: lookup ap-argocd.dsp-devops.broadinstitute.org on 169.254.169.254:53: read udp 172.17.0.2:44840->169.254.169.254:53: i/o timeout"
```
https://fc-jenkins.dsp-techops.broadinstitute.org/job/fiab-start/87003/console

File: README.md
Comment (tlangs) [inline README.md]: if you're running M1, wouldn't you want the `linux/arm64` mysql image?

File: http/src/test/scala/org/broadinstitute/dsde/workbench/leonardo/config/ConfigSpec.scala
Comment (tlangs) [inline http/src/test/scala/org/broadinstitute/dsde/workbench/leonardo/config/ConfigSpec.scala:118]: If we do this now, Leo tests will talk to DRSHub, but deployed Leo will talk to Martha. Do we want that?

Comment (andy7i) [reply http/src/test/scala/org/broadinstitute/dsde/workbench/leonardo/config/ConfigSpec.scala:118]: yeah, I also had that question: https://github.com/DataBiosphere/leonardo/pull/2997#discussion_r1024532347

File: README.md
Comment (andy7i) [reply README.md]: running m1, we need to force pull the linux/x86_64 image because the linux/arm64 one does not exist for 5.6
Screen Shot 2022-11-17 at 12 23 27 PM

I'll make that more explicit!

File: http/src/test/scala/org/broadinstitute/dsde/workbench/leonardo/config/ConfigSpec.scala
Comment (andy7i) [reply http/src/test/scala/org/broadinstitute/dsde/workbench/leonardo/config/ConfigSpec.scala:118]: also, do you know the answer to this question? https://github.com/DataBiosphere/leonardo/pull/2997#discussion_r1024537436

File: README.md
Comment (andy7i) [reply README.md]: I guess the other way to solve this is to add the docker pull command to https://github.com/DataBiosphere/leonardo/blob/develop/docker/run-mysql.sh

Maybe I'll pull this out into a separate PR.

Comment (andy7i) [reply README.md]: pulling this into a new PR https://github.com/DataBiosphere/leonardo/pull/3023

**Comment by andy7i** [comment]
jenkins retest
```
11:21:25 4:21PM ERR 500 Internal Server Error from https://firecloudorch.jenkins-swat-leo-5476.bee.envs-terra.bio/register/profile ({"causes":[],"message":"The http server closed the connection unexpectedly before delivering responses for 1 outstanding requests","source":"FireCloud","stackTrace":[]})
```
https://fc-jenkins.dsp-techops.broadinstitute.org/job/fiab-start/87054/console

**Comment by andy7i** [comment]
jenkins retest - different failure each time. unstable tests?

**Comment by gpolumbo-broad** [comment]
When we worked with Justin on Thursday to do the independisam release, he was commenting that the Leo tests were having all sorts of problems.  Not sure what they were or if they're ongoing or if they affect only the main branch or what.  But I know Leo tests were struggling.

**Comment by Qi77Qi** [comment]
@gpolumbo-broad if you're referring to Justin Yang, I think he's probably talking about when running in github workflow, which has its own problems...Regarding tests, there isn't known consistent failures...what I've seen myself is lots of fiab-start errors

File: project/Settings.scala
Comment (Qi77Qi) [inline project/Settings.scala]: this needs to be reverted

File: http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/config/Config.scala
Comment (Qi77Qi) [reply http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/config/Config.scala]: for just tests, just use whatever that's in reference.conf

File: http/src/main/resources/reference.conf
Comment (tlangs) [reply http/src/main/resources/reference.conf]: They should be equivalent, as DRSHub is a direct port of Martha.

Comment (Qi77Qi) [reply http/src/main/resources/reference.conf]: cool..once this is deployed..can we also make sure things are working as expected in Galaxy on dev?

cc @afgane

File: http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/config/Config.scala
Comment (andy7i) [reply http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/config/Config.scala]: just to clarify, you're saying we should keep the old values (referencing Martha), correct?

If so, I'll leave this PR open until the cutover to DRSHub.

File: http/src/main/resources/reference.conf
Comment (andy7i) [reply http/src/main/resources/reference.conf]: yeah, how would we do that? Would these be manual tests / are they written down somewhere?

Comment (Qi77Qi) [reply http/src/main/resources/reference.conf]: I don't actually know very well how it's being used either...was hoping @afgane could provide some more info

Comment (andy7i) [reply http/src/main/resources/reference.conf]: ok this PR is more of a cleanup PR so I'm going to put this under [ID-347](https://broadworkbench.atlassian.net/browse/ID-347) instead. 

@afgane please let us know how to test galaxy either here, slack (#dsp-identiteam) or on https://broadworkbench.atlassian.net/browse/ID-313

**Comment by rtitle** [comment]
Just checking back on this PR, seems it's not doing much anymore -- just changing the default `drsUrl` in reference.conf which is overridden anyway by firecloud-develop. 

So I'm happy we landed on a simpler solution to just configure the DRS URL environmentally -- am OK with merging or closing this PR.

File: http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/config/Config.scala
Comment (Qi77Qi) [reply http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/config/Config.scala]: what I'm saying is for tests, you can just update it to whatever value you change to in `reference.conf`...like here in this PR, you updated to drs url, so just use that in test too...

The real value is controlled by config value in firecloud-develop, so whether or not we want to merge this PR is up-to-you...doesn't really matter whether we merge now or later as long as `firecloud-develop` has correct value for each env

Comment (andy7i) [reply http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/config/Config.scala]: oh gotcha, thanks for clarifying! I'll go ahead and merge this then to reduce code hanging around

**Comment by andy7i** [comment]
@Qi77Qi do you mind reviewing this PR in the next week or two?

**Comment by Qi77Qi** [comment]
I thumbed your merging comment https://github.com/DataBiosphere/leonardo/pull/2997#discussion_r1030552008  ...guess I  could explicitly approve if u want

**Review by Qi77Qi — APPROVED** [review]

**Comment by andy7i** [comment]
> I thumbed your merging comment [#2997 (comment)](https://github.com/DataBiosphere/leonardo/pull/2997#discussion_r1030552008) ...guess I could explicitly approve if u want

Thank you!

**Pull request merged**
